core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java

```diff
@@ -297,31 +292,38 @@ private void abortL1UpdateOrWait(Object key) throws InterruptedException {
          }
          return ctx.continueInvocation();
       }
-      Future<?> l1InvalidationFuture = invalidateL1(ctx, command, assumeOriginKeptEntryInL1);
-      Object returnValue = ctx.forkInvocationSync(command);
-      processInvalidationResult(command, l1InvalidationFuture);
-      removeFromLocalL1(ctx, command);
-      return ctx.shortCircuit(returnValue);
+      Future<?> l1InvalidationFuture = invalidateL1InCluster(ctx, command, assumeOriginKeptEntryInL1);
+      return ctx.forkInvocation(command, (ctx1, command1, rv, throwable) -> {
```

Comment:
This is one of the subtler limitations of the async interceptors... `ReturnHandler`s are not allowed to call `forkInvocation`, only `ForkReturnHandler`s are. In other words, you cannot call `forkInvocation()` after you called `onReturn()`. And because `removeFromLocalL1()` calls `forkInvocation()`, the interceptor can't use `onReturn()`.

The `ReturnHandler`/`ForkReturnHandler` distinction is quite annoying at times, but on the positive side helps that you can call `onReturn` multiple times in the same `visit*Command` method, and it works like nested `try/finally` blocks. Also, `onReturn` is a bit cheaper :)